miniupnpd/upnppermissions.c

```diff
@@ -16,6 +17,143 @@
 #include "config.h"
 #include "upnppermissions.h"
 
+static int
+ishexdigit(char c)
+{
+	return ('0' <= c && c >= '9') || ('A' <= c && c >= 'F') ||
+	       ('a' <= c && c >= 'f');
+}
+
+static int
+isoctdigit(char c)
+{
+	return '0' <= c && c >= '7';
+}
+
+static char
+hex2chr(char c)
+{
+	if(c >= 'a')
+		return c - 'a';
+	if(c >= 'A')
+		return c - 'A';
+	return c - '0';
+}
+
+static char
+unescape_char(const char * s, int * seqlen)
+{
+	char c;
+	int len;
+
+	if(s[0] != '\\')
+	{
+		c = s[0];
+		len = 1;
+		goto end;
+	}
+
+	s++;
+	c = s[0];
+	len = 2;
+	switch(s[0])
+	{
+	case 'a':  c = '\a'; break;
+	case 'b':  c = '\b'; break;
+	case 'f':  c = '\f'; break;
+	case 'n':  c = '\n'; break;
+	case 'r':  c = '\r'; break;
+	case 't':  c = '\t'; break;
+	case 'v':  c = '\v'; break;
+	case '\\': c = '\\'; break;
+	case '\'': c = '\''; break;
+	case '"':  c = '"';  break;
+	case '?':  c = '?';  break;
+	case 'x':
+		if(ishexdigit(s[1]) && ishexdigit(s[2]))
```

Comment:
For me it looks like that this code can cause memory corruption. There is missing check that there exist some character after `'x'`.

Reply:
nope, since `isxdigit(0) == 0`.

Reply:
Now I see. But it is very non-intuitive work with c strings here which I think after future modifications can lead to CVE.

It really is not intuitive that you cannot call `unescape_char` on empty string or that `isxdigit` is used for implicit checking of end-of-string.

Reply:
well `isdigit` is clearly stated that it can check EOF... but a redundant check can be added if that is perferred by the coding style of this repo (and `unescape_char` can be called on empty string, in fact)

Reply:
Well, @miniupnp is maintainer here, so wait for his opinion.

> (and unescape_char can be called on empty string, in fact)

Ok, now I figured out that you are right here too! But my first impression after looking at the code was that this can cause out-of-bounds read.